design/ceph/osd-migration.md

```diff
@@ -0,0 +1,207 @@
+- To support any new and improved storage backend, the existing OSD needs to be destroyed, disks should be wiped out and new OSD should be created using the latest storage backend.
+
+## GOALS
+- Newly provisioned clusters should be able to use the specific storage backend while creating OSDs.
+- Existing clusters should be able to migrate the OSDs to use a specific storage backend, without downtime and without risking data loss.
+
+
+## PROPOSAL
+- Add a new field called `spec.Storage.osdStore` in the ceph cluster yaml.
+- This will contain:
+  - `spec.Storage.osdStore.type`: The type of backend store to be used for OSDs. For example, bluestore, bluestore_rdr, etc. The default store type will be `bluestore`
+  - `spec.Storage.osdStore.updateStore`: Whether the user wants to migrate the existing OSDs to use new storage backend. The migration process will involve destroying the existing OSDs one by one, wiping the disks, creating a new OSD and waiting for the PGs to be active+clean before migrating the next OSD. Since this operation involves possible downtime, users should be really sure before proceeding with this action. This field can only take the value as `yes-really-update-store`
+  -
+
+- Example:
+```yaml
+storage:
+  osdStore:
+    type: bluestore_rdr
+    updateStore: yes-really-update-store
+```
+
+- New deployments will use this `osdStore.type` while creating the OSDs. If empty, then `bluestore` will be used by default.
+- If the user wants to change the `osdStore.type` field for an existing cluster, then they will also need to update `spec.Storage.osdStore.updateStore` with `yes-really-update-store` in order to migrate the existing osds to use updated osdStore.
+
+*NOTE*:
+- Once the OSDs are migrated to a new OSD store, say, `bluestore-rdr``, they won't be allowed to be migrated back to the legacy store (bluestore)
+
+- Add a new label `osd-store:<osd store type>` to all the OSD pods.
+- This label will help to identify the current backend store being used for the OSD and enable the operator to decide if the
+  OSD should be migrated in case the user changes the backend store type in the spec (`spec.Storage.osdStore.type`).
+
+
+### NEW CLUSTERS
+- Currently the osd backend store is hardcoded to `bluestore` throughout the operator code.
+- Add a new environment variable called `ROOK_OSD_STORE_TYPE` in the OSD prepare and OSD pods. This will contain the OSD store type. Use this environment variable to prepare and activate the cluster.
+
+#### OSD Prepare
+- The `ceph-volume prepare` command is updated to use backend store provided in the `spec.storage.osdStore`. If no value is provided for `spec.storage.osdStore` then default to `bluestore` backend.
+
+  - **OSD on disks**:
```

Comment:
OSDs on PVC?

Reply:
The commands are same for both OSD on PVC and OSD on disks.  So removed the reference to `OSD on disks` here.